File: 8_heap/ModifiableHeap.ts
type Token<T> = {
  value: T
  heapIndex: number
}

/**
 * 可修改堆.
 * @deprecated 比`ErasableHeap`慢.
 */
class ModifiableHeap<T> {
  private readonly _data: Token<T>[]
  private readonly _less: (a: T, b: T) => boolean

  constructor({ data, less }: { data: T[]; less: (a: T, b: T) => boolean }) {
    this._data = data.map((value, heapIndex) => ({ value, heapIndex }))
    this._less = less
    if (this._data.length > 1) this._heapify()
  }

  push(value: T): Token<T> {
    const res = { value, heapIndex: this._data.length }
    this._data.push(res)
    this._up(this._data.length - 1)
    return res
  }

  pop(): Token<T> {
    if (!this._data.length) throw new Error('pop from an empty heap')
    const n = this._data.length - 1
    this._swap(0, n)
    this._down(0, n)
    return this._data.pop()!
  }

  top(): Token<T> {
    if (!this._data.length) throw new Error('top from an empty heap')
    return this._data[0]
  }

  remove(token: Token<T>): void {
    this._remove(token.heapIndex)
  }

  modify(token: Token<T>, value: T): void {
    token.value = value
    this._fix(token.heapIndex)
  }

  clear(): void {
    this._data.length = 0
  }

  get size(): number {
    return this._data.length
  }

  private _heapify(): void {
    const n = this._data.length
    for (let i = (n >>> 1) - 1; ~i; i--) {
      this._down(i, n)
    }
  }

  private _up(j: number): void {
    const { _data, _less } = this
    while (j) {
      const i = (j - 1) >>> 1 // parent
      if (i === j || !_less(_data[j].value, _data[i].value)) break
      this._swap(i, j)
      j = i
    }
  }

  private _down(i0: number, n: number): boolean {
    const { _data, _less } = this
    let i = i0
    while (true) {
      const j1 = (i << 1) | 1
      if (j1 >= n || j1 < 0) break
      let j = j1
      const j2 = j1 + 1
      if (j2 < n && _less(_data[j2].value, _data[j1].value)) j = j2
      if (!_less(_data[j].value, _data[i].value)) break
      this._swap(i, j)
      i = j
    }
    return i > i0
  }

  private _fix(i: number): void {
    if (!this._down(i, this._data.length)) this._up(i)
  }

  private _remove(i: number): Token<T> {
    const n = this._data.length - 1
    if (n !== i) {
      this._swap(i, n)
      if (!this._down(i, n)) this._up(i)
    }
    return this._data.pop()!
  }

  private _swap(i: number, j: number): void {
    const tmp = this._data[i]
    this._data[i] = this._data[j]
    this._data[j] = tmp
    this._data[i].heapIndex = i
    this._data[j].heapIndex = j
  }
}

export { ModifiableHeap }

if (require.main === module) {
  const pq = new ModifiableHeap<number>({ data: [1, 8, 3, 5], less: (a, b) => a < b })
  console.log(pq.top())

  {
    const token = pq.push(4)
    console.log(pq.pop())
    console.log(pq.pop())
    pq.remove(token)
    console.log(pq.pop())
    console.log(pq.pop())
  }

  {
    const token = pq.push(-1)
    pq.push(111)
    pq.push(1)
    pq.modify(token, 100)
    console.log(pq.pop())
    console.log(pq.pop())
    console.log(pq.pop())
  }

  console.log(pq.size)

  const N = 1e7
  // test perf
  console.time('ErasableHeap')

  const pq2 = new ModifiableHeap<number>({ data: [], less: (a, b) => a < b })
  const tokens = Array<Token<number>>(N)
  for (let i = 0; i < N; ++i) {
    tokens[i] = pq2.push(i)
    pq2.top()
  }
  for (let i = 0; i < N; ++i) {
    pq2.remove(tokens[i])
  }
  console.timeEnd('ErasableHeap') // 1e7: 1.34s
}

File: 8_heap/modifiableHeap/ModifiableHeap.go
// https://github.com/EndlessCheng/codeforces-go/blob/master/copypasta/heap.go

package main

import (
	"container/heap"
	"fmt"
)

func main() {
	mh := NewModifiableHeap[int](0, func(a, b int) bool { return a < b })

	{
		token := mh.Push(1)
		mh.Push(2)
		mh.Push(3)
		mh.Modify(token, 4)
		fmt.Println(mh.Pop().value) // 2
		fmt.Println(mh.Pop().value) // 3
		fmt.Println(mh.Pop().value) // 4
	}

	{
		mh.Push(3)
		token := mh.Push(2)
		mh.Remove(token)
		fmt.Println(mh.Top().value) // 3
	}

}

type ModifiableHeap[T any] struct{ wrapper *pq[T] }

// 支持修改、删除指定元素的堆
// 用法：调用 push 会返回一个 *viPair 指针，记作 p
// 将 p 存于他处（如 slice 或 map），可直接在外部修改 p.v 后调用 fix(p.hi)，从而做到修改堆中指定元素
// heap.Remove(p.hi) 通过 swap_remove 实现，复杂度 O(logn).
func NewModifiableHeap[T any](initCapacity int32, less func(a, b T) bool) *ModifiableHeap[T] {
	if initCapacity < 0 {
		initCapacity = 0
	}
	wrapper := &pq[T]{make([]*viPair[T], 0, initCapacity), less}
	return &ModifiableHeap[T]{wrapper}
}

func (h *ModifiableHeap[T]) Push(v T) *viPair[T] {
	p := &viPair[T]{v, int32(len(h.wrapper.data))}
	heap.Push(h.wrapper, p)
	return p
}
func (h *ModifiableHeap[T]) Pop() *viPair[T]     { return heap.Pop(h.wrapper).(*viPair[T]) }
func (h *ModifiableHeap[T]) Top() *viPair[T]     { return h.wrapper.data[0] }
func (h *ModifiableHeap[T]) Len() int32          { return int32(len(h.wrapper.data)) }
func (h *ModifiableHeap[T]) Init()               { heap.Init(h.wrapper) }
func (h *ModifiableHeap[T]) Remove(p *viPair[T]) { heap.Remove(h.wrapper, int(p.heapIndex)) }
func (h *ModifiableHeap[T]) Modify(p *viPair[T], v T) {
	p.value = v
	heap.Fix(h.wrapper, int(p.heapIndex))
}
func (h *ModifiableHeap[T]) Clear() {
	h.wrapper.Clear()
}

var _ heap.Interface = &pq[any]{}

type viPair[T any] struct {
	value     T
	heapIndex int32
}

type pq[T any] struct {
	data []*viPair[T]
	less func(a, b T) bool
}

func (w *pq[T]) Len() int           { return len(w.data) }
func (w *pq[T]) Less(i, j int) bool { return w.less(w.data[i].value, w.data[j].value) }
func (w *pq[T]) Swap(i, j int) {
	data := w.data
	data[i], data[j] = data[j], data[i]
	data[i].heapIndex = int32(i)
	data[j].heapIndex = int32(j)
}
func (w *pq[T]) Push(v any) { w.data = append(w.data, v.(*viPair[T])) }
func (w *pq[T]) Pop() any {
	res := w.data[len(w.data)-1]
	w.data = w.data[:len(w.data)-1]
	return res
}
func (w *pq[T]) Clear() {
	w.data = w.data[:0]
}
